File: leetcode/1-100/41-first-missing-positive/bench/first_missing_positive.c
// LeetCode #41 bench mirror — C, the in-place cyclic-sort solver (★).
//
// Mirrors bench/first_missing_positive.kara: swap each in-range value to its home index
// v-1, then scan for the first slot not holding its home value. The buffer is allocated
// once and overwritten in place each iteration. Same workload (TOTAL refills of a k-rotated
// permutation with a punched gap) and the same answer-fold checksum as every other mirror.

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

static int64_t first_missing_positive(int64_t *nums, int64_t n) {
    int64_t i = 0;
    while (i < n) {
        int64_t v = nums[i];
        if (v >= 1 && v <= n && nums[v - 1] != v) {
            int64_t t = nums[v - 1];
            nums[v - 1] = v;
            nums[i] = t;
        } else {
            i++;
        }
    }
    for (int64_t j = 0; j < n; j++) {
        if (nums[j] != j + 1) {
            return j + 1;
        }
    }
    return n + 1;
}

int main(void) {
    const int64_t total = 200000;
    const int64_t n = 100;
    const int64_t modulus = 1000000007;

    int64_t *nums = (int64_t *)malloc((size_t)n * sizeof(int64_t));

    int64_t acc = 0;
    for (int64_t k = 0; k < total; k++) {
        int64_t rot = k % n;
        for (int64_t i = 0; i < n; i++) {
            nums[i] = ((i + rot) % n) + 1;
        }
        nums[k % n] = n + 7;

        int64_t ans = first_missing_positive(nums, n);
        acc = (acc * 131 + ans) % modulus;
    }

    free(nums);
    printf("%lld\n", (long long)acc);
    return 0;
}

File: leetcode/1-100/41-first-missing-positive/bench/first_missing_positive.rs
// LeetCode #41 bench mirror — Rust, the in-place cyclic-sort solver (★).
//
// Mirrors bench/first_missing_positive.kara exactly: swap each in-range value to its home
// index v-1, then scan for the first slot not holding its home value. The buffer is
// allocated once and overwritten in place each iteration with a k-rotated permutation of
// 1..N plus a punched-out gap; fold each answer into a rolling checksum. Same workload +
// sink as every other mirror.

fn first_missing_positive(nums: &mut [i64], n: i64) -> i64 {
    let mut i = 0i64;
    while i < n {
        let v = nums[i as usize];
        if v >= 1 && v <= n && nums[(v - 1) as usize] != v {
            let t = nums[(v - 1) as usize];
            nums[(v - 1) as usize] = v;
            nums[i as usize] = t;
        } else {
            i += 1;
        }
    }
    let mut j = 0i64;
    while j < n {
        if nums[j as usize] != j + 1 {
            return j + 1;
        }
        j += 1;
    }
    n + 1
}

fn main() {
    let total: i64 = 200000;
    let n: i64 = 100;
    let modulus: i64 = 1000000007;

    let mut nums: Vec<i64> = vec![0; n as usize];

    let mut acc: i64 = 0;
    let mut k: i64 = 0;
    while k < total {
        let rot = k % n;
        let mut i = 0i64;
        while i < n {
            nums[i as usize] = ((i + rot) % n) + 1;
            i += 1;
        }
        nums[(k % n) as usize] = n + 7;

        let ans = first_missing_positive(&mut nums, n);
        acc = (acc * 131 + ans) % modulus;
        k += 1;
    }

    println!("{}", acc);
}
